### ISSUE TITLE
How to check the profit in a trade before it ends

### ISSUE THREAD

**Issue by claudioivp** [Lu-Yi-Hsun/iqoptionapi#133]
Hi, congratulations by your api

When I use a Martingale's strategy is important know the status of my trading 3 seconds before it ends to start a new trade for the begining of the next candlestick. How is the best approach to do this? Unfortunatelly the method check_win_digital_v2 in a While condition is not enough because the next trading starts with too many seconds late.

Big hug.

**Comment by Lu-Yi-Hsun** [comment]
So you want  check win speed up?

**Comment by claudioivp** [comment]
exactly, to prevent any delay before the next trading starts at the begining next candlestick

**Comment by Lu-Yi-Hsun** [comment]
@claudioivp 

this sample can check win or loss how much money in hight speed

```python
from iqoptionapi.stable_api import IQ_Option
import logging
import time
#logging.basicConfig(level=logging.DEBUG,format='%(asctime)s %(message)s')
I_want_money=IQ_Option("email","password")
 
ACTIVES="EURUSD"
duration=1#minute 1 or 5
amount=1
action="call"#put

print("__For_Binary_Option__")
_,id=I_want_money.buy(amount,ACTIVES,action,duration)
while I_want_money.get_async_order(id)==None:
    pass

while I_want_money.get_async_order(id)["name"]!="option-closed":
    pass
close=I_want_money.get_async_order(id)
print(close)
print("get money")
print(close["msg"]["win_enrolled_amount"]-close["msg"]["enrolled_amount"])
 
```

**Comment by claudioivp** [comment]
Hello, good morning!
Sorry, I'm not a python programmer, so I'm not understood behavior of this implementation well.
See it running here:
https://www.youtube.com/watch?v=wH46t3Tt1G0&feature=youtu.be

**Comment by claudioivp** [comment]
Please ignore my last message, my api version was outdated. Now i'm checking your code.

**Comment by claudioivp** [comment]
Very nice, dear friend.
Could you turn on this implementation for Digital Options too?

**Comment by Lu-Yi-Hsun** [comment]
@claudioivp 
```python
from iqoptionapi.stable_api import IQ_Option
import logging
import time
#logging.basicConfig(level=logging.DEBUG,format='%(asctime)s %(message)s')
I_want_money=IQ_Option("email","password")
 
ACTIVES="EURUSD"
duration=1#minute 1 or 5
amount=1
action="call"#put

print("__For_Digital_Option_Spot_")
id=I_want_money.buy_digital_spot(ACTIVES,amount,action,duration)
 
while True:
    check,win=I_want_money.check_win_digital_v2(id)
    if check:
        break
print(I_want_money.check_win_digital_v2(id))
```

**Comment by claudioivp** [comment]
Hi there,

hmmmmmm, fine, now I understand, but we need come back some steps...
This approach works very well in options, but not in digital, let me explain

My initial tests weremade in digital, and as I said at the begining this issue, "check_win_digital_v2 in a While condition is not enough because the next trading starts with too many seconds late" 

For some reason, using your last sample you can check what I'm saying, your api need some seconds to get the trading result, and then  persistent trading (martingale) starts some seconds late.

See this video to you understand what I'm talking about:
https://youtu.be/N4jsq-d97RQ

So this is the reason for my suggestion, someway, to verify the status (profit) of my trading 3 seconds before it ends to start a new trade for the begining of the next candlestick.

Big hug, dear friend

**Comment by Lu-Yi-Hsun** [comment]
@claudioivp  
In the last version check_win_digital(),implement another way.

But in "check_win_digital()" and "check_win_digital_v2()" still have about 3 sec delay because of IQoption Server problem.

```python
from iqoptionapi.stable_api import IQ_Option
import logging
import time
#logging.basicConfig(level=logging.DEBUG,format='%(asctime)s %(message)s')
I_want_money=IQ_Option("email","password")
 
ACTIVES="EURUSD-OTC"
duration=1#minute 1 or 5
amount=1
action="call"#put
from datetime import datetime
 
id=I_want_money.buy_digital_spot(ACTIVES,amount,action,duration) 

while True:
    check,_=I_want_money.check_win_digital(id)
    if check:
        break
print(I_want_money.get_digital_position(id))
print(I_want_money.check_win_digital(id))
```

**Comment by claudioivp** [comment]
Fine, but and about the idea that I have proposed, about check status of the trading 3 seconds before it closes (or any time 0:xy) in a open negotiation, even if the trading has not been finalised? This is not possible?

**Comment by Lu-Yi-Hsun** [comment]
https://github.com/Lu-Yi-Hsun/iqoptionapi/blob/master/README.md#get_async_order

**Comment by claudioivp** [comment]
Hi there, I try this, but when checking the json generated, field pnl_realized_enrolled has not been seted yet. And there is no current price in any field. So I'm not have an attribute to check or compare. Do you have some idea what could I do... ?

**Comment by Lu-Yi-Hsun** [comment]
What current price ?

**Comment by claudioivp** [comment]
I mean that I don't have any parameter to know if my trading is having loss or win when using get_async_order. at least I heaven't found one

**Comment by Lu-Yi-Hsun** [comment]
https://github.com/Lu-Yi-Hsun/iqoptionapi#check_win_digital_v2

check_win_digital_v2 is implement by get_async_order

you can read the source code:
https://github.com/Lu-Yi-Hsun/iqoptionapi/blob/master/iqoptionapi/stable_api.py#L852

**Comment by samdevbr** [comment]
The class below can work with martingales using this repository, it's a part of my trading bot.

`class OrderMonitor(Thread):
    QUOTE_STEP = 1
    ORDER_STEP = 2
    FINISH_STEP = 3

    symbol = None
    step = QUOTE_STEP
    time = None
    action = None
    won = False

    position_value = 350
    order_price = 0
    gales = 0
    max_gales = 1

    def __init__(self, symbol: str, time: datetime, action: str, max_gales: int = 1):
        Thread.__init__(self)

        self.symbol = symbol
        self.time = time
        self.action = action
        self.max_gales = max_gales

        self.set_position_value()

    def set_position_value(self):
        pass

    def open_position(self, current_price):
        if self.gales > 0:
            self.position_value *= 2

        self.order_price = current_price
        context.buy_digital_spot(self.symbol, self.position_value, self.action, IQ_ORDER_EXPIRATION_M1)
        sleep(1.5)

    def winning(self, price):
        return price > self.order_price if self.action == "call" else price < self.order_price

    def check_order(self, quotes):
        _, candle = quotes[0]

        now = IQGetClock(context) + timedelta(seconds=1) # I_want_money.get_server_timestamp()
        closing_time = datetime.fromtimestamp(candle['to'])
        current_price = Decimal(candle['close'])

        winning = self.winning(current_price)

        if now.timestamp() == closing_time.timestamp() and winning:
            print("%s won - o: %d c: %d\n" % (self.symbol, self.order_price, current_price))
            self.step = self.FINISH_STEP
            self.won = True
        elif now.timestamp() == closing_time.timestamp() and not winning and self.gales < self.max_gales:
            self.gales += 1
            print("%s %s gale %i\n" % (self.symbol, self.time.time(), self.gales))
            self.open_position(Decimal(current_price))
        elif now.timestamp() == closing_time.timestamp() and self.gales == self.max_gales:
            print("%s oops" % self.symbol)
            self.step = self.FINISH_STEP

    def check_quote(self, quotes):
        _, candle = quotes[0]

        now = IQGetClock(context) + timedelta(seconds=1) # I_want_money.get_server_timestamp()

        if now.timestamp() == self.time.timestamp():
            self.open_position(Decimal(candle['close']))
            self.step = self.ORDER_STEP

    def run(self):
        cprint("\nOpening quote pool for %s" % self.symbol, "white", "on_blue")
        IQOpenQuotePool(context, self.symbol) # I_want_money.start_candles_stream()
        quote_pool = IQReceiveQuote(context, self.symbol) # I_want_money.get_realtime_candles()
        passed = self.time < datetime.now()

        if not passed:
            while self.step != self.FINISH_STEP:
                quotes = list(quote_pool)

                if self.step == self.QUOTE_STEP:
                    self.check_quote(quotes)
                elif self.step == self.ORDER_STEP:
                    self.check_order(quotes)
`

**Comment by JafferWilson** [comment]
@samdevbr What is the issue then please let us now?
